lib/puppet/application/lookup.rb

```diff
@@ -361,17 +361,22 @@ def generate_scope
     unless node.is_a?(Puppet::Node) # to allow unit tests to pass a node instance
       ni = Puppet::Node.indirection
       tc = ni.terminus_class
-      if tc == :plain || options[:compile]
-        node = ni.find(node, facts: extra_facts)
-      else
-        ni.terminus_class = :plain
-        node = ni.find(node, facts: extra_facts)
-        ni.terminus_class = tc
-      end
+      temp_node = Puppet::Node.new(options[:node])
+      temp_node.parameters['clientcert'] = options[:node]
+      overrides = {
+          trusted_information: extra_facts.values['trusted']
+        }
+        Puppet.override(overrides) do
```

Comment:
Ideally I think we want to construct a "full" trusted context in the same way that the compiler would do. Unfortunately, the compiler and CA may be different hosts, so we can't just look in the local filesystem for the  agent's cert. But we could query the CA for the cert:

```ruby
service = Puppet.runtime[:http].route_to(:ca)
cert = service.get_certificate(node)
trusted = Puppet::Context::TrustedInformation(true, node, cert)
Puppet.override(trusted_information: trusted) do
  ...
end
```

Then the classifier terminus will call [`to_h`](https://github.com/puppetlabs/classifier/blob/6e007c5760896c0beb6cf31cf6768e729f02370c/puppet/lib/puppet/indirector/node/classifier.rb#L74) on that.

One important thing is that if puppetserver is configured to use a source of trusted *external* facts, then those facts will also be added to the trusted info (and can be used for lookup classification) https://github.com/puppetlabs/puppet/blob/8b4331b9f5b4e548621b76796f082770de1778d9/lib/puppet/context/trusted_information.rb#L48